sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -362,7 +362,11 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] with PredicateHelper {
         if (topNPushed) {
           sHolder.pushedLimit = Some(limit)
           sHolder.sortOrders = orders
-          operation
+          sHolder.builder match {
+            case s: SupportsPushDownTopN if s.supportCompleteSortPushDown() =>
+              operation
+            case _ => s
```

Comment:
This makes the top N pushdown look like a hint: we notify the source that a top N is needed, but the Spark query plan is not changed.

This is definitely not optimal: Spark still does a sort, and probably a shuffle if the limit is large. Performance-wise, I think it's the same as pushing down a limit. How about a simpler approach: if we can't push down top N, we just push down limit without changing the plan. Then we don't need to change DS v2 APIs.